mpeg2/src/interleaving_muxer.rs

```diff
@@ -0,0 +1,463 @@
+            }
+        }
+        Ok(())
+    }
+
+    fn write_muxer_packet(&mut self, stream_index: usize, mut p: Packet) -> Result<(), EncodeError> {
+        if let Some(ref mut pdt) = p.dts_90khz {
+            *pdt &= TS_33BIT_MASK;
+        } else if let Some(ref mut pts) = p.pts_90khz {
+            *pts &= TS_33BIT_MASK;
+        }
+        let stream = &mut self.streams[stream_index].inner;
+        self.inner.write(stream, p)?;
+        Ok(())
+    }
+
+    fn write_packets_outside_buffer_duration(&mut self) -> Result<(), EncodeError> {
+        let mut packets = vec![];
+        for (stream_index, stream) in self.streams.iter_mut().enumerate() {
+            while let Some(p) = stream.buffered_packets.front() {
+                let ts = p.dts_90khz.or(p.pts_90khz).unwrap_or(stream.last_written_ts);
+                if self.largest_ts_in_buffer - ts > self.max_buffer_duration_90khz {
+                    packets.push((
+                        stream_index,
+                        ts,
+                        stream.buffered_packets.pop_front().expect("there must be at least one packet to pop"),
+                    ));
+                } else {
+                    break;
+                }
+            }
+        }
+        for (stream_index, ts, packet) in packets {
+            self.write_muxer_packet(stream_index, packet)?;
+            self.streams[stream_index].last_written_ts = ts;
+        }
+        Ok(())
+    }
+
+    fn min_ts_excluding(&self, stream_id: usize) -> u64 {
+        let mut min_ts = u64::MAX;
+        for (_, stream) in self.streams.iter().enumerate().filter(|&(index, _)| index != stream_id) {
+            if let Some(p) = stream.buffered_packets.front() {
+                min_ts = min_ts.min(p.dts_90khz.or(p.pts_90khz).unwrap_or(stream.last_written_ts));
+            } else {
+                min_ts = 0;
+                break;
+            }
+        }
+        min_ts
+    }
+
+    pub fn flush(&mut self, ignore_error: bool) -> Result<(), EncodeError> {
+        self.largest_ts_in_buffer = 0;
+
+        let mut packets = vec![];
+        self.streams.iter_mut().enumerate().for_each(|(stream_index, stream)| {
+            while let Some(p) = stream.buffered_packets.pop_front() {
+                packets.push((stream_index, stream.last_written_ts, p));
+            }
+            stream.last_written_ts = 0;
+        });
+
+        packets.sort_unstable_by_key(|(_, ts, p)| p.dts_90khz.or(p.pts_90khz).unwrap_or(*ts));
+        for (stream_index, _, packet) in packets {
+            match self.write_muxer_packet(stream_index, packet) {
+                Err(e) if !ignore_error => {
+                    return Err(e);
+                }
+                _ => {}
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<W: Write> Drop for InterleavingMuxer<W> {
+    fn drop(&mut self) {
+        self.flush(true).expect("Errors are ignored when flushing the buffered packets");
```

Comment:
I would recommend getting rid of the `ignore_error` argument and just ignoring the errors like this:

```suggestion
        // Errors are ignored when flushing the buffered packets.
        let _ = self.flush();
```

It makes `flush` simpler, and if `flush` encounters an error, it should return immediately instead of continuing to try to write the other packets.

Reply:
Agree!